### PR TITLE
Fix GPX tracking restarts

### PR DESCRIPTION
**Pull request by dg0yt**

### CODE REVIEW

**Comment by dg0yt** [comment]
I couldn't test this yet. 

It is not urgent. There is a workaround, and there won't be new Android builds until 0.8.1.

**Comment by lpechacek** [comment]
Bug replication procedure:
* Start Mapper with a large map and templates to make it allocate memory
* Start position recording
* Put Mapper into background and start another memory hungry application which will evict Mapper from memory
* Restart Mapper and activate Position recording again - _at this point the previous GPX track segment was lost from the on-disk .gpx file_

This bug bit me a few times. I'll give the fix a shot ASAP.

**Comment by dg0yt** [comment]
IMO you don't need to force a crash, cf #1055.

**Comment by lpechacek** [comment]
> IMO you don't need to force a crash, cf #1055.

IIRC when I stopped recording and closed the application proper, no data loss occurred. My memory may be weak though....

File: src/gui/map/map_editor.cpp [hunk 1]
```diff
@@ -3332,37 +3332,47 @@ void MapEditorController::enableGPSDisplay(bool enable)
...
-					if (map->getTemplate(i)->getTemplateState() != Template::Loaded)
+					if (qstrcmp(temp->getTemplateType(), "TemplateTrack") == 0)
 					{
-						// If the template file could not be loaded, don't care
-						// at this point; simply create a new file.
-						TemplateTrack* track = static_cast<TemplateTrack*>(map->getTemplate(i));
-						track->configureForGPSTrack();
-						track->setHasUnsavedChanges(true);
+						// Re-use this TemplateTrack.
+						track = static_cast<TemplateTrack*>(temp);
```
[the remark this answers is not in the document]
Comment (dg0yt) [reply src/gui/map/map_editor.cpp]: I considered `qobject_cast`, too, but I didn't want to change too much in master for now.  

The whole template family of classes needs a refactoring. I will probably start with that when I work on GDAL raster image support, est. between v0.9 and v1.0. `TemplateTrack` can be eventually removed if we manage to handle tracking with OGR template class.

**Comment by dg0yt** [comment]
> >    IMO you don't need to force a crash, cf #1055.
>
> IIRC when I stopped recording and closed the application proper, no data loss occurred. My memory may be weak though....

It is created by a new combination of circumstances, such as making GDAL/OGR the default, and no longer loading the template data on construction. That's what I hadn't understood in the first version of my PR: It is no longer enough to construct the `TemplateTrack` for the existing file but we also need to explicitly load the data.

**Comment by ollesmaps** [comment]
Maybe this relates. On two ocassions when Android shut down (probably I pushed turnoff button by accident), I got a new gpx track when Mapper restarted (and location turned on) with the same name as the gpx before shut down. So there were two gpxs with the same name in templates. Only the newer one could be seen on map. It was not posssible to show on map the gpx of the older one.
(That was with Mapper version one before 0.8.)

**Comment by dg0yt** [comment]
The old code creates a new template when the existing template has the wrong type (GDAL/OGR) for tracking (TemplateTrack). And the existing template would not notice the changes in the file until the map is reopened. 
This is fixed by this proposal. too: The existing template is removed, and the new template is inserted at the same place. The only thing which I didn't care for is saving the visibility attributes of the existing template.

**Comment by dg0yt** [comment]
This seems to work for me. Okay to merge?

**Comment by lpechacek** [comment]
I haven't tested yet but the code change look pretty good to me. I'm fine with merge.

**Comment by lpechacek** [comment]
Tested yesterday, works as expected. Data is no longer lost and track segments are added as expected. Thanks for fixing this annoying bug, Kai!

**Pull request merged**